### PR TITLE
[POC][tests-only][full-ci] run e2e tests with runner

### PR DESCRIPTION
**Pull request by saw-jan**
## Description
To better organize the growing e2e tests, we have to move features to their favorable suites (folders). Also, we have to be careful about the execution time of each e2e pipeline. With the current approach and feature file location it will be difficult to organize and manage the pipeline execution time.

Therefore, here is my proposal:
1. Move feature files into different suites (folders) and maintain this location: `features/<suite>/*.feature`
    - move folders/suites out of `smoke` suite
    ```
    features/smoke/shares -> features/shares
    features/smoke/search -> features/search
    ...
    ```
2. Use runner script to run the e2e tests
    - doesn't affect the current way of running e2e tests
    - organizing suites in CI is simplified
    ```diff
    -"featurePaths": [
    -    "tests/e2e/cucumber/features/{smoke,journeys}/*.feature",
    -]
    +"suites": [
    +    "journeys",
    +    "smoke",
    +]
    ```
    - ability to split and run existing suites into groups (with this ocis ci doesn't have to change anything if some suites are added to the web repo)

#### Run all suites
```
./run-e2e.sh
```
#### Run test providing paths
```
./run-e2e.sh cucumber/features/journeys cucumber/features/search/search.feature
```
#### Run single scenario
```
./run-e2e.sh cucumber/features/search/search.feature:122
```
#### Run selective suites
```
./run-e2e.sh --suites shares,spaces
```
#### Exclude suites
```
./run-e2e.sh --xsuites app-provider,search
```
#### Divide and run suites into parts (suitable for splitting e2e in oCIS ci)
```
./run-e2e.sh --run-part 2 --total-parts 4
```
```
./run-e2e.sh --run-part 1 --total-parts 4 --exclude app-provider
```

#### Help
```bash
$ ./run-e2e.sh -h

COMMAND [options] [paths]

Available options:
    --suites        - suites to run. Comma separated values (folder names)
                      e.g.: --suites smoke,shares
    --xsuites       - exclude suites from running. Comma separated values
                      e.g.: --exclude spaces,search
    --run-part      - part to run out of total parts (groups)
                      e.g.: --run-part 2 (runs part 2 out of 4)
    --total-parts   - total number of groups to divide into
                      e.g.: --total-parts 4 (suites will be divided into 4 groups)
    --help, -h      - show cli options
```

## Motivation and Context
Make it easier to add/skip/organize e2e test suites in a pipeline

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests
- [ ] Documentation
- [ ] Maintenance (e.g. dependency updates or tooling)

### CODE REVIEW

File: .drone.star [hunk 1]
```diff
@@ -62,29 +62,32 @@ config = {
...
+                "search",
+                "shares",
             ],
         },
         "oCIS-app-provider": {
             "skip": False,
-            "featurePaths": [
-                "tests/e2e/cucumber/features/smoke/app-provider/*.feature",
+            "suites": [
+                "app-provider",
```
Comment (saw-jan) [inline .drone.star:90]: not working for single suite. have to fix it

**Comment by ScharfViktor** [comment]
Wow, looks really good. 

one remark about path: 

image

I guess `run-e2e.sh` file should be located in the `web`?  if we use `--feature tests/e2e/cucumber/features` ?

or 

`cd tests/e2e`
`run-e2e.sh --feature cucumber/features/smo...`

I wanted to suggest you use Makefile, but it's only for CI. localy we can still to use `pnpm test:e2e:cucumber tes..`

**Comment by saw-jan** [comment]
>I guess run-e2e.sh file should be located in the web? if we use --feature tests/e2e/cucumber/features ?

currently, it should `test/e2e/...`/ but yeah, good catch :+1: . since the runner is in the tests/e2e, the path should be relative to it.

>I wanted to suggest you use Makefile, but it's only for CI. localy we can still to use pnpm test:e2e:cucumber tes..

yeah, we still can use that command

File: .drone.star (hunk 1 above)
Comment (saw-jan) [reply .drone.star:90]: fixed: https://github.com/owncloud/web/pull/10557/commits/635a478805467d7186a26a86b225c64ae5d1aa0d

**Comment by saw-jan** [comment]
>>I guess run-e2e.sh file should be located in the web? if we use --feature tests/e2e/cucumber/features ?
>currently, it should test/e2e/.../ but yeah, good catch 👍 . since the runner is in the tests/e2e, the path should be relative to it.

Fixed! Now paths should start from `cucumber/...`, relative to script path

**Comment by ScharfViktor** [comment]
```
$ ./run-e2e.sh -h

COMMAND [options] [paths]

Available options:
    --suites        - suites to run. Comma separated values (folder names)
                      e.g.: --suites smoke,shares
    --exclude       - exclude suites from running. Comma separated values
                      e.g.: --exclude spaces,search
    --run-part      - part to run out of total parts (groups)
                      e.g.: --run-part 2 (runs part 2 out of 4)
    --total-parts   - total number of groups to divide into
                      e.g.: --total-parts 4 (suites will be divided into 4 groups)
    --help, -h      - show cli options
```

I see that now we have 7 suites: `admin-settings`, `app-provider`, `jorneys`, `search`, `shares`, `smoke`, `spaces`

1. if we have folder with test in the suite-> scripts runs only tests in the folder
Screenshot 2024-03-07 at 11 38 07

2.  `--exclude` Maybe we should use exclude for `features`? like a `./run-e2e.sh --suites admin-settings --exclude general.feature` Because if you choose which suites should run piplene -> you don't exclude any suites (but excluding feature would be great IMHO)

3. regarding `--run-part` `--total-parts`  do we need it?

**Comment by ScharfViktor** [comment]
> Yeah, I am planning to use those in ocis CI. because with the current implementation if we add new folder in web, ocis will miss that. (we can also use them in web ci as well)

but we can't just split tests and run them. many tests require a specific configuration (tika app-provider) or use like `./run-e2e.sh --run-part 2 --total-parts 4 --exclude app-provider,search`

**Comment by saw-jan** [comment]
>but we can't just split tests and run them. many tests require a specific configuration (tika app-provider) or use like ./run-e2e.sh --run-part 2 --total-parts 4 --exclude app-provider,search

right (both) :+1: . I am thinking of the same for special suites `./run-e2e.sh --run-part 2 --total-parts 4 --exclude app-provider,search`. then add them is separate pipelines

**Comment by saw-jan** [comment]
>>--exclude Maybe we should use exclude for features?

>Great! I will try to find out something for that one too

@ScharfViktor For now, I want to provide an option to exclude suites only. Because implementing feature file exclusion option needs evaluating all feature files available (within the execution context). I vote for implementing it only when feel the urge for it.

What would you suggest?

Also, change in one option:
`--exclude` -> `--xsuites`

so that if we want to implement feature file exclusion option then we can do `--xfeatures`

**Comment by ScharfViktor** [comment]
> What would you suggest?

Yeah, let's decide that when it's really necessary.

**Review by ScharfViktor — APPROVED** [review]

**Pull request merged**